[Doc] Add documentions for layout (#900)
taichi-dev/taichi: docs/snode.rst

        ::

            snode.dense(ti.ijk, 3)

        will be translated into:

        ::

            snode.dense(ti.ijk, (3, 3, 3))

.. function:: snode.dynamic(index, size, chunk_size = None)

    :parameter snode: (SNode) parent node where the child is derived from
    :parameter index: (Index) index used for this node
    :parameter size: (scalar) the maximum size of the dynamic node
    :parameter chunk_size: (optional, scalar) (TODO: describe this)
    :return: (SNode) the derived child node

    The size of dynamic nodes can be extended in runtime, see functions below.

    This places a 1-D dynamic tensor of maximal size ``12``:

    ::

        ti.root.dynamic(ti.i, 12).place(x)

.. function:: ti.length(snode)

    :parameter snode: (SNode, dynamic)
    :return: (scalar) current size of the dynamic node

.. function:: ti.append(snode, indices, val)

    :parameter snode: (SNode, dynamic)
    :parameter indices: (scalar or tuple) indices within SNode
    :parameter val: (depends on SNode data type) value to store

>     ASK(yuanming-hu): how is this used exactly??

Author: Hello?

> 
>     :parameter snode: (SNode) parent node where the child is derived from
>     :parameter index: (Index) index used for this node
>     :parameter size: (scalar) the maximum size of the dynamic node
>     :parameter chunk_size: (optional, scalar) (TODO: describe this)
>     :return: (SNode) the derived child node
> 
>     The size of dynamic nodes can be extended in runtime, see functions below.
> 
>     This places a 1-D dynamic tensor of maximal size ``12``:
> 
>     ::
> 
>         ti.root.dynamic(ti.i, 12).place(x)
> 
> 
> .. function:: ti.length(snode)
> 
>     :parameter snode: (SNode, dynamic)
>     :return: (scalar) current size of the dynamic node
> 
> 
> .. function:: ti.append(snode, indices, val)
> 
>     :parameter snode: (SNode, dynamic)
>     :parameter indices: (scalar or tuple) indices within SNode
>     :parameter val: (depends on SNode data type) value to store
> 
>     ASK(yuanming-hu): how is this used exactly??
> 
> 

Author: ```suggestion
```

.. _snode:

Structural nodes (SNodes)
=========================

After writing the computation code, the user needs to specify the internal data structure hierarchy. Specifying a data structure includes choices at both the macro level, dictating how the data structure components nest with each other and the way they represent sparsity, and the micro level, dictating how data are grouped together (e.g. structure of arrays vs. array of structures).
Our language provides *structural nodes (SNodes)* to compose the hierarchy and particular properties. These constructs and their semantics are listed below:

* dense: A fixed-length contiguous array.

> * bitmasked: Use a mask to maintain sparsity information, one bit per child.

Reviewer: ```suggestion
* bitmasked: This is similar to dense, but it also uses a mask to maintain sparsity information, one bit per child.
```

* pointer: Store pointers instead of the whole structure to save memory and maintain sparsity.

> * hash: Use a hash table to maintain the mapping from active coordinates to data addresses in memory. Suitable for high sparsity.

Reviewer: I don't think `hash` is available in Py Taichi yet..?

Our language provides *structural nodes* to compose the hierarchy and particular properties. These constructs and their semantics are listed below:

* dense: A fixed-length contiguous array.

* dynamic: Variable-length array, with a predefined maximum length. It serves the role of ``std::vector``, and can be used to maintain objects (e.g. particles) contained in a block.

* bitmasked: Use a mask to maintain sparsity infomation, one bit per child.

* pointer: Store pointers instead of the whole structure to save memory and maintain sparsity.

* hash: Use a hash table to maintain the mapping from active coordinates to data addresses in memory. Suitable for high sparsity.

See :ref:`layout` for more details about data layout.

.. function:: snode.place(x, ...)

    :parameter snode: (SNode) where to place
    :parameter x: (tensor) tensor(s) to be placed
    :return: (SNode) the ``snode`` itself

> 

Author: ```suggestion

.. function:: snode.dim()

    :parameter snode: (SNode or tensor)
    :return: (scalar) the dimension of node / tensor
    
    Equivalent to len(snode.shape)
    
    ::
        
        ti.root.dense(ti.ij, (3, 4)).place(x)
        x.dim()  # 2
```
